aggregator_api/src/lib.rs

```diff
@@ -0,0 +1,897 @@
+        Arc::clone(conn.state().expect("No AggregatorApi in connection state"));
+
+    conn = aggregator_api.check_auth(conn);
+    if !conn.is_halted() {
+        conn = handler(Arc::clone(&aggregator_api), conn).await;
+    }
+
+    let status: u16 = match conn.status() {
+        Some(status) => status.into(),
+        None => {
+            warn!(path = conn.path(), "No status set on connection");
+            return conn;
+        }
+    };
+
+    aggregator_api.response_time_histogram.record(
+        &Context::current(),
+        start_instant.elapsed().as_secs_f64(),
+        &[
+            KeyValue::new("endpoint", endpoint_name.into()),
+            KeyValue::new("status", status.to_string()),
+        ],
+    );
+    conn
+}
+
+struct AggregatorApi<C: Clock> {
+    ds: Datastore<C>,
+    cfg: Config,
+    response_time_histogram: Histogram<f64>,
+}
+
+impl<C: Clock> AggregatorApi<C> {
+    fn check_auth(&self, conn: Conn) -> Conn {
+        if let Some(authorization_value) = conn.headers().get("authorization") {
+            if let Some(received_token) = authorization_value.as_ref().strip_prefix(b"Basic ") {
+                if self.cfg.auth_tokens.iter().any(|key| {
+                    constant_time::verify_slices_are_equal(received_token, key.as_ref()).is_ok()
+                }) {
+                    // Authorization succeeds.
+                    return conn;
+                }
+            }
+        }
+
+        // Authorization fails.
+        conn.with_status(Status::Unauthorized).halt()
+    }
+
+    async fn handle_get_task_ids(&self, conn: Conn) -> Conn {
+        const PAGINATION_TOKEN_KEY: &str = "pagination_token";
+        let lower_bound = querify(conn.querystring())
+            .into_iter()
+            .find(|&(k, _)| k == PAGINATION_TOKEN_KEY)
+            .map(|(_, v)| TaskId::from_str(v))
+            .transpose();
+        let lower_bound = match lower_bound {
+            Ok(task_id_param) => task_id_param,
+            Err(_) => return conn.with_status(Status::BadRequest).halt(),
+        };
+
+        let task_ids = conn_try!(
+            self.ds
+                .run_tx_with_name("get_task_ids", |tx| Box::pin(async move {
+                    tx.get_task_ids(lower_bound).await
+                }))
+                .await,
+            conn
+        );
+        let pagination_token = task_ids.last().cloned();
+
+        conn.with_json(&GetTaskIdsResp {
+            task_ids,
+            pagination_token,
+        })
+    }
+
+    async fn handle_post_task(&self, mut conn: Conn) -> Conn {
+        let req: PostTaskReq = match conn.deserialize().await {
+            Ok(req) => req,
+            Err(_) => return conn.with_status(Status::BadRequest).halt(),
+        };
+
+        let vdaf_verify_keys = Vec::from([SecretBytes::new(
+            thread_rng()
+                .sample_iter(Standard)
+                .take(req.vdaf.verify_key_length())
+                .collect(),
+        )]);
```

Comment:
This implementation assumes that it's our responsibility to generate a task ID and a verify key. I think we need to support being handed those values in the request, or we won't be able to test this flow with two Januses.

Reply:
You're right, though this API as-written is "just" for the initial "create and sync tasks between Janus Control & Data Plane" problem, rather than the harder "sync task from ourselves to a co-aggregator" or "have task sync'ed to us from our co-aggregator" problems. These cross-organization syncs will likely need to allow (almost) every parameter to be specified directly, and I think we'd want to define the appropriate API in the Control Plane first (since that's what the co-aggregators would be talking to).

WDYT about waiting to implement this?

Reply:
Sounds good to me!